Removes packages.config file in NuGet.Client

dominoFire opened the pull request:
## Bug

Fixes: https://github.com/NuGet/Home/issues/8630
Regression: No
* Last working version: [NuGet.Client@dc72141d](https://github.com/NuGet/NuGet.Client/commit/dc72141d53cdec46b13c5a4a08ffdd5cefe7a3a1)
* How are we preventing it in future: By dogfooding our own features in NuGet.Client repo

## Fix

- Removes .nuget\packages.config file from the repo
- Removes `nuget.exe` dependencies on local build scenarios and in some CI build steps
- Makes necessary changes to bootstrap packages from build\build.proj when needed

## Testing/Validation

Tests Added: No  
Reason for not adding tests: Infrastructure/Engineering changes
Validation: CI build

## Notes

Special thanks to @zivkan for his great previous work on this PR

zivkan committed:
Create project file to restore repo bootstrap files

zivkan committed:
References to bootstrapped nuget packages migrated to v3 folder structure

zivkan committed:
Remove .nuget/packages.config and configure scripts to use msbuild restore

zivkan committed:
work

Commit:
Updated end2end test data creation

Commit:
Corrected package version in e2e test data generation

Commit:
Bootstrap target

- Added BootstrapPackages to build.proj
- Replaced APEX task for bootstraping packages
- Fixed package path in xplat functional tests

Commit:
reverted build.proj

Commit:
Diagnostic E2E tests on DDNuGetW001

Commit:
Diagnostic MacOS 10.14 and DDNuGetW

Commit:
Removed fixed agent names

nkolev92 commented on build/common.ps1:
```diff
@@ -349,7 +331,7 @@ Function Clear-PackageCache {
     Trace-Log 'Cleaning local caches'
 
-    & nuget locals all -clear -verbosity detailed
+    & dotnet nuget locals all --clear --verbosity detailed
```
this is illegal. 

dotnet nuget locals doesn't have a verbosity switch.

dominoFire replied:
Removed flag.

nkolev92 commented on global.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "msbuild-sdks": {
+        "Microsoft.Build.NoTargets": "1.0.80"
```
Can we put the version in the bootstrap file itself?

dominoFire replied:
Removed global.json and added directly to build\build.proj

nkolev92 reviewed (commented):
A couple of remarks. 

Looks good other than that.

Commit:
Addressed feedback

- Removed global.json
- corrected dotnet cleaning nuget cache
- specific version in bootstrap.proj

nkolev92 commented on build/bootstrap.proj:
```diff
@@ -0,0 +1,31 @@
+    <PropertyGroup>
+        <RestorePackagesPath>$(MSBuildThisFileDirectory)../packages/</RestorePackagesPath>
+        <BaseIntermediateOutputPath>$(RestorePackagesPath)obj/</BaseIntermediateOutputPath>
+        <IntermediateOutputPath>$(BaseIntermediateOutputPath)$(Configuration)/</IntermediateOutputPath>
```
Why do we need this?

dominoFire replied:
It's an attempt to move the generated obj folder outside of `$NUGET_ROOT\build\` folder

nkolev92 replied:
This is a  no build SDK, so only NuGet is the one that creates assets, and NuGet reads MSBuildProjectExtensionsPath

dominoFire replied:
Ok, I will use `$(MSBuildThisFileDirectory)` instead of `$(BaseIntermediateOutputPath)`

nkolev92 commented on build/bootstrap.proj:
```diff
@@ -0,0 +1,31 @@
+<Project Sdk="Microsoft.Build.NoTargets/1.0.80">
+    <PropertyGroup>
+        <RestorePackagesPath>$(MSBuildThisFileDirectory)../packages/</RestorePackagesPath>
+        <BaseIntermediateOutputPath>$(RestorePackagesPath)obj/</BaseIntermediateOutputPath>
```
We should use MSBuildProjectExtensionsPath instead of BaseIntermediateOutputPath. 
Right now I'd imagine it's like that the assets file ends up in some unexpected place.

dominoFire replied:
Sounds good but, I would need to import the props and targets of the project. Let me see how it works.

nkolev92 replied:
> I would need to import the props and targets of the proj

Not sure what you mean by this.

dominoFire replied:
I mean adding import statements to load `\build\common.props` and `"$(BuildCommonDirectory)common.targets` from this repo (NuGet.Client)

nkolev92 replied:
Isn't it the other way around?


Specifically the following should be enough.

```
    <PropertyGroup>
        <RestorePackagesPath>$(MSBuildThisFileDirectory)../packages/</RestorePackagesPath>
        <MSBuildProjectExtensionsPath>$(RestorePackagesPath)obj/</MSBuildProjectExtensionsPath>
        <TargetFramework>netstandard2.0</TargetFramework>
        <DisableImplicitNuGetFallbackFolder>true</DisableImplicitNuGetFallbackFolder>
    </PropertyGroup>
```

dominoFire replied:
Oh. I didn't know that. Let me test this change

dominoFire replied:
It works perfectly. I suggest to move the project.assets.json of this bootstrap.proj to `$NUGET_ROOT\artifacts\packages\obj\project.assets.json` to avoid mixing this file with the `$NUGET_ROOT\packages\` folder

nkolev92 replied:
I'm ok with that. 
The caches and assets file are only used when configure.ps1 is invoked so the footprint should be negligible.

zivkan replied:
FYI I intentionally did not put the assets file in the `artifacts` directory because some build steps like signing and localisation takes all assemblies in the `artifacts` directory. Now, this bootstrap project doesn't create assemblies, so it's potentially not a problem, but in general test projects don't put their assemblies in `artifacts` either. 

Since the `artifacts` folder appears to be mostly for projects that get shipped to customers, it didn't feel appropriate to me to put the bootstrap assets file there.

nkolev92 commented on build/bootstrap.proj:
```diff
@@ -0,0 +1,31 @@
+        <PackageDownload Include="Microsoft.VisualStudio.Composition" version="[15.8.98]" />
+        <PackageDownload Include="Microsoft.VisualStudio.ProjectSystem" version="[16.0.201-pre-g7d366164d0]" />
+        <PackageDownload Include="Microsoft.VSSDK.BuildTools" version="[15.9.3032]" />
+        <PackageDownload Include="Microsoft.Web.Xdt" version="[2.1.2]" />
```
Now that this is a not packages.config anymore we can consider a having the versions in common.props or common.project.props or something that. 
Specifically this applies for packages that are both in the bootstrap and as a PackageReference. 

So Web.Xdt, Newtonsoft.Json are definitely some of those. 

Then again, you can also do all of them, then we'll never need to change any of the includes such as

 `<ILMergeExePath>$(SolutionPackagesFolder)ilmerge\3.0.21\tools\net452\ILMerge.exe</ILMergeExePath>`

nkolev92 replied:
Reviewing further down, we have the exact path of the assemblies in the packages, so maybe it doesn't matter as much :)

dominoFire replied:
Some bootstrap packages need some more adjustments when they are referenced in other projects. For example:

https://github.com/NuGet/NuGet.Client/blob/840b09a4cecda6ae8fd1c07967874fe8061ece63/src/NuGet.Clients/NuGet.VisualStudio.Client/NuGet.VisualStudio.Client.csproj#L71

With this PR, the path would change to:

```
 <Content Include="$(SolutionPackagesFolder)Lucene.Net\3.0.3\lib\NET40\Lucene.Net.dll"> 
```

nkolev92 replied:
yeah, that's what I meant in the 2nd comment. 

I don't feel too strongly either way, so make a judgment call

nkolev92 commented:
@dominoFire why was the progress on this halted?

Commit:
Changes property to MSBuildProjectExtensionsPath

zivkan commented:
@dominoFire Can you rebase and resolve the merge conflict? It looks like the only reason we couldn't merge this a month ago was because of a test infrastructure problem. Hopefully we can complete it this time.

nkolev92 commented:
As he's on vacation, I took the liberty to rebase his PR.
